### PR TITLE
stats: Fix gRPC xDS pending_requests: now drains back to 0

### PR DESCRIPTION
**Pull request by fredlas**
*Description*: xDS gRPC's stat for number of xDS request queued to be sent was previously only getting set when an attempt to send hit rate limiting. In addition to not updating when new items arrive, this stat would never return to 0 once set to non-0.

*Risk Level*: low
*Testing*: added coverage of fixed thing to existing unit test

### CODE REVIEW

**Comment by fredlas** [comment]
Two notes: 1) I intend to move checkRateLimitAllowsDrain() to private; not doing it yet to keep the more important changes easy to see in the diff, and 2) I don't understand why the test isn't passing. I think I either don't exactly understand what the test harness is doing, or stats_store_.gauge().value() does not refer to the same stats store I'm causing a change to.

**Comment by htuch** [comment]
@fredlas can you look into the build failures? Seems legit. Feel free to make the helper private as needed.

**Comment by fredlas** [comment]
Turns out I was applying the test framework correctly, and just not understanding the logic being tested! I now realize that the request queue *only* gets (meaningfully) used in cases of rate limiting. Left a comment to clarify.

**Comment by fredlas** [comment]
After working with the guts of stats in another PR, I realized why this one was originally using set() like it was: it was avoiding activating this stat in the (common) case that rate limiting was not active. I switched from inc/dec back to set. But, the set() is now in a different place, where it will be correct.

File: source/common/config/grpc_stream.h [hunk 1]
```diff
@@ -109,11 +96,26 @@ class GrpcStream : public Grpc::TypedAsyncStreamCallbacks<ResponseProto>,
...
   void drainRequests() {
     ENVOY_LOG(trace, "draining discovery requests {}", request_queue_.size());
-    while (!request_queue_.empty() && checkRateLimitAllowsDrain(request_queue_.size())) {
+    while (!request_queue_.empty() && checkRateLimitAllowsDrain()) {
       // Process the request, if rate limiting is not enabled at all or if it is under rate limit.
       sendDiscoveryRequest(request_queue_.front());
       request_queue_.pop();
     }
+    if (request_queue_.size() > 0 || control_plane_stats_.pending_requests_.used()) {
+      control_plane_stats_.pending_requests_.set(request_queue_.size());
```
[the remark this answers is not in the document]
Comment (fredlas) [reply source/common/config/grpc_stream.h]: Done; I tried to cover the answer to your question in the added comment.

**Comment by htuch** [comment]
Please merge master to pick up https://github.com/envoyproxy/envoy/pull/5827.

**Comment by fredlas** [comment]
Merged, thanks for the heads up.

**Comment by fredlas** [comment]
Ping; should be ready to go now.

File: source/common/config/grpc_stream.h [hunk 2]
```diff
@@ -109,11 +96,33 @@ class GrpcStream : public Grpc::TypedAsyncStreamCallbacks<ResponseProto>,
...
+    while (!request_queue_.empty() && checkRateLimitAllowsDrain()) {
       // Process the request, if rate limiting is not enabled at all or if it is under rate limit.
       sendDiscoveryRequest(request_queue_.front());
       request_queue_.pop();
     }
+    // Although request_queue_.push() happens elsewhere, the only time the queue is non-transiently
+    // non-empty is when it remains non-empty after a drain attempt. (The push() doesn't matter
+    // because we always attempt this drain immediately after the push). Basically, a change in
+    // queue length is not "meaningful" until it has persisted until here. Since we are constantly
+    // doing set(0), we need the if(>0 || used) to keep this stat from being wrongly marked
```
Comment (htuch) [inline source/common/config/grpc_stream.h]: What do you mean by "constantly doing set(0)"? Where does that happen?

File: source/common/config/grpc_stream.h [hunk 3]
```diff
@@ -109,11 +96,33 @@ class GrpcStream : public Grpc::TypedAsyncStreamCallbacks<ResponseProto>,
...
+    // interesting and needlessly taking up space. The first time we set(123), used becomes true,
```
Comment (htuch) [inline source/common/config/grpc_stream.h]: I don't see how the `set()` makes `used` true below... we can chat IRL tomorrow if that's easier.

File: source/common/config/grpc_stream.h (hunk 2 above)
Comment (fredlas) [reply source/common/config/grpc_stream.h]: Every DiscoveryRequest goes through this queue+drain process, even if it's not actually going to be queued. I did write it kind of weird, didn't I: I really meant "if not for this conditional we would be doing it constantly." Clarified the comment.

File: source/common/config/grpc_stream.h (hunk 3 above)
Comment (fredlas) [reply source/common/config/grpc_stream.h]: Stats have used() false until they are first given a value; for a gauge that includes set(), even set(0).  https://github.com/envoyproxy/envoy/blob/master/include/envoy/stats/stats.h#L58    and implementation at   https://github.com/envoyproxy/envoy/blob/master/source/common/stats/stat_data_allocator_impl.h#L134

**Comment by mattklein123** [comment]
@fredlas I don't see any test change in the diff? Did a change get dropped accidentally?

**Comment by htuch** [comment]
@mattklein123 I see comment changes in https://github.com/envoyproxy/envoy/pull/5796/commits/7998d8daa4db07e474599bc60118edd4d4891c52, which covers my last set of Qs.

**Comment by mattklein123** [comment]
@htuch Shouldn't this change have a test that failed before the fix? I don't see any test changes?

**Comment by fredlas** [comment]
Whoops, yeah. I started with tests that were testing what I (wrongly) thought the logic was supposed to be, and removed them once I realized I was testing the wrong thing, but didn't add anything else. #5903 tests the fixed behavior. Sorry!

**Pull request merged**